src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

```diff
@@ -0,0 +1,368 @@
+            }
+        },
+        EAGER_GLOBAL_ORDINALS {
+            @Override
+            public String toString() {
+                return EAGER_GLOBAL_ORDINALS_VALUE;
+            }
+        };
+
+        public static final String KEY = "loading";
+        public static final String EAGER_GLOBAL_ORDINALS_VALUE = "eager_global_ordinals";
+        public static final String EAGER_VALUE = "eager";
+        public static final String LAZY_VALUE = "lazy";
+
+        public static Loading parse(String loading, Loading defaultValue) {
+            if (Strings.isNullOrEmpty(loading)) {
+                return defaultValue;
+            } else if (EAGER_GLOBAL_ORDINALS_VALUE.equalsIgnoreCase(loading)) {
+                return EAGER_GLOBAL_ORDINALS;
+            } else if (EAGER_VALUE.equalsIgnoreCase(loading)) {
+                return EAGER;
+            } else if (LAZY_VALUE.equalsIgnoreCase(loading)) {
+                return LAZY;
+            } else {
+                throw new MapperParsingException("Unknown [" + KEY + "] value: [" + loading + "]");
+            }
+        }
+    }
+
+    private Names names;
+    private float boost;
+    // TODO: remove this docvalues flag and use docValuesType
+    private boolean docValues;
+    private NamedAnalyzer indexAnalyzer;
+    private NamedAnalyzer searchAnalyzer;
+    private NamedAnalyzer searchQuoteAnalyzer;
+    private SimilarityProvider similarity;
+    private Loading normsLoading;
+    private FieldDataType fieldDataType;
+
+    protected MappedFieldType(MappedFieldType ref) {
+        super(ref);
+        this.names = ref.names();
+        this.boost = ref.boost();
+        this.docValues = ref.hasDocValues();
+        this.indexAnalyzer = ref.indexAnalyzer();
+        this.searchAnalyzer = ref.searchAnalyzer();
+        this.searchQuoteAnalyzer = ref.searchQuoteAnalyzer();
+        this.similarity = ref.similarity();
+        this.normsLoading = ref.normsLoading();
+        this.fieldDataType = ref.fieldDataType();
+    }
+
+    public MappedFieldType() {}
+
+    public MappedFieldType clone() {
+        return new MappedFieldType(this);
+    }
+
+    public boolean isNumeric() {
+        return false;
+    }
+
+    public boolean isSortable() {
+        return true;
+    }
+
+    public Names names() {
+        return names;
+    }
+
+    public void setNames(Names names) {
+        checkIfFrozen();
+        this.names = names;
+    }
+
+    public float boost() {
+        return boost;
+    }
+
+    public void setBoost(float boost) {
+        checkIfFrozen();
+        this.boost = boost;
+    }
+
+    public FieldDataType fieldDataType() {
+        return fieldDataType;
+    }
+
+    public void setFieldDataType(FieldDataType fieldDataType) {
+        checkIfFrozen();
+        this.fieldDataType = fieldDataType;
+    }
+
+    public boolean hasDocValues() {
+        return docValues;
+    }
+
+    public void setHasDocValues(boolean hasDocValues) {
+        checkIfFrozen();
```

Comment:
The call to super for AbstractFieldMapper calls freeze, so it is already guaranteed that the type is frozen at construction time. But a test would be good for this.

I've added a norelease comment to override freeze() to validate all settings are set. I'll do this in a follow up because keeping in sync with master has already been painful after only a couple days.